### PR TITLE
Merging ModelBinding into Mvc.Core

### PR DESCRIPTION
**Pull request by rynowak**

### CODE REVIEW

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/PrefixContainer.cs [hunk 1]
```diff
@@ -3,6 +3,7 @@
 
 using System;
 using System.Collections.Generic;
+using System.Diagnostics;
 using Microsoft.Framework.Internal;
 
 namespace Microsoft.AspNet.Mvc.ModelBinding.Internal
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/PrefixContainer.cs]: Could you move this to a `Internal` subdirectory \ remove the `Internal` namespace?

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/PrefixContainer.cs [hunk 2]
```diff
@@ -187,6 +188,36 @@ internal static bool IsPrefixMatch(string prefix, string testString)
             }
         }
 
+        /// <summary>
+        /// Convert an ICollection to an array, removing null values. Fast path for case where
+        /// there are no null values.
+        /// </summary>
+        private static T[] ToArrayWithoutNulls<T>(ICollection<T> collection) where T : class
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/PrefixContainer.cs]: `TVal`

**Comment by Eilon** [comment]
Also very nice...

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/PrefixContainer.cs (hunk 2 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/PrefixContainer.cs]: renaming to `TElement` which we use in general in the collection model binders.

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/PrefixContainer.cs (hunk 1 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/PrefixContainer.cs]: Making this public and not in the internal namespace. You'd need this to implement a value provider correctly.

**Pull request closed without merging**